OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/OpenPgpServiceKeyIdExtractor.java

```diff
@@ -53,150 +47,223 @@ private OpenPgpServiceKeyIdExtractor(ContentResolver contentResolver,
+    private KeyIdResult returnKeyIdsFromEmails(Intent data, String[] encryptionUserIds, String callingPackageName) {
         boolean hasUserIds = (encryptionUserIds != null && encryptionUserIds.length > 0);
 
+        boolean anyKeyNotVerified = false;
         HashSet<Long> keyIds = new HashSet<>();
         ArrayList<String> missingEmails = new ArrayList<>();
         ArrayList<String> duplicateEmails = new ArrayList<>();
+        HashMap<String,KeyRow> keyRows = new HashMap<>();
         if (hasUserIds) {
-            for (String rawUserId : encryptionUserIds) {
-                OpenPgpUtils.UserId userId = KeyRing.splitUserId(rawUserId);
-                String email = userId.email != null ? userId.email : rawUserId;
-                // try to find the key for this specific email
-                Uri uri = KeyRings.buildUnifiedKeyRingsFindByEmailUri(email);
-                Cursor cursor = contentResolver.query(uri, KEY_SEARCH_PROJECTION, KEY_SEARCH_WHERE, null, null);
-                if (cursor == null) {
-                    throw new IllegalStateException("Internal error, received null cursor!");
-                }
-                try {
-                    // result should be one entry containing the key id
-                    if (cursor.moveToFirst()) {
-                        long id = cursor.getLong(INDEX_MASTER_KEY_ID);
-                        keyIds.add(id);
-
-                        // another entry for this email -> two keys with the same email inside user id
-                        if (!cursor.isLast()) {
-                            Log.d(Constants.TAG, "more than one user id with the same email");
-                            duplicateEmails.add(email);
-
-                            // also pre-select
-                            while (cursor.moveToNext()) {
-                                long duplicateId = cursor.getLong(INDEX_MASTER_KEY_ID);
-                                keyIds.add(duplicateId);
-                            }
-                        }
-                    } else {
-                        missingEmails.add(email);
-                        Log.d(Constants.TAG, "user id missing");
+            Uri queryUri = EmailStatus.CONTENT_URI.buildUpon().appendPath(callingPackageName).build();
+            Cursor cursor = contentResolver.query(queryUri, PROJECTION_KEY_SEARCH, null, encryptionUserIds, null);
+            if (cursor == null) {
+                throw new IllegalStateException("Internal error, received null cursor!");
+            }
+
+            try {
+                while (cursor.moveToNext()) {
+                    String queryAddress = cursor.getString(INDEX_EMAIL_ADDRESS);
+                    Long masterKeyId = cursor.isNull(INDEX_MASTER_KEY_ID) ? null : cursor.getLong(INDEX_MASTER_KEY_ID);
+                    int verified = cursor.getInt(INDEX_EMAIL_STATUS);
+
+                    KeyRow row = new KeyRow(masterKeyId, verified == 2);
```

Comment:
To make my comment more clear ;) It is difficult to see if the new code in ``OpenPgpServiceKeyIdExtractor`` has the provides the same functionality and works for the same corner cases as before.